include/boost/gil/image_processing/numeric.hpp

```diff
@@ -295,6 +295,120 @@ inline void compute_hessian_entries(
     detail::convolve_2d(dy, sobel_y, ddyy);
 }
 
+namespace detail {
+enum class direction: std::size_t {
+    north = 0,
+    south = 1,
+    west = 2,
+    east = 3,
+    north_east = 4,
+    south_east = 5,
+    south_west = 6,
+    north_west = 7
+};
+
+template <typename OutputView, typename InputView>
+void compute_nabla(InputView view, const std::vector<OutputView>& nabla) {
+    constexpr std::ptrdiff_t input_num_channels = num_channels<InputView>{};
+    static_assert(num_channels<OutputView>{} == input_num_channels, "input and output views must have the same amount of channels");
+    for (std::ptrdiff_t y = 1; y < view.height() - 1; ++y)
+    {
+        for (std::ptrdiff_t x = 1; x < view.width() - 1; ++x)
+        {
+            for (std::ptrdiff_t channel_index = 0; channel_index < input_num_channels; ++channel_index)
+            {
+                nabla[(std::size_t)direction::north](x, y)[channel_index] = view(x, y - 1)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::south](x, y)[channel_index] = view(x, y + 1)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::west](x, y)[channel_index] = view(x - 1, y)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::east](x, y)[channel_index] = view(x + 1, y)[channel_index] - view(x, y)[channel_index];
+
+                nabla[(std::size_t)direction::north_east](x, y)[channel_index] = view(x + 1, y - 1)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::south_east](x, y)[channel_index] = view(x + 1, y + 1)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::south_west](x, y)[channel_index] = view(x - 1, y + 1)[channel_index] - view(x, y)[channel_index];
+                nabla[(std::size_t)direction::north_west](x, y)[channel_index] = view(x - 1, y - 1)[channel_index] - view(x, y)[channel_index];
+            }
+        }
+    }
+}
+
+template <typename View>
+void calculate_diffusvity(std::vector<View> nablas, double kappa, const std::vector<View> diffusivities)
+{
+    using pixel_type = typename View::value_type;
+    using channel_type = typename channel_type<View>::type;
+    BOOST_ASSERT(nablas.size() == diffusivities.size());
```

Comment:
The Why not an exception?

A while ago we had lengthy discussion on assertions vs exceptions for user-specified input, see #360, as well as on Slack [here](https://cpplang.slack.com/archives/C27KZLB0X/p1564768229242000) from Peter 

> you should never assert on input
> input is not a programming error, because it's not the programmer who has supplied it